File: Sources/APTradeApplication/NewsUseCases.swift
import APTradeDomain

/// Fetches market headlines for a category. Best-effort: network/rate-limit failures
/// resolve to an empty list (the view model distinguishes no-key from empty separately).
public struct FetchMarketNewsUseCase: Sendable {
    private let repository: NewsRepository
    public init(repository: NewsRepository) { self.repository = repository }
    public func callAsFunction(category: NewsCategory) async -> [NewsArticle] {
        (try? await repository.marketNews(category: category)) ?? []
    }
}

/// Fetches recent company news for a symbol. Best-effort, like `FetchMarketNewsUseCase`.
public struct FetchCompanyNewsUseCase: Sendable {
    private let repository: NewsRepository
    public init(repository: NewsRepository) { self.repository = repository }
    public func callAsFunction(symbol: String) async -> [NewsArticle] {
        (try? await repository.companyNews(symbol: symbol)) ?? []
    }
}

public struct LoadBookmarksUseCase: Sendable {
    private let store: BookmarkStore
    public init(store: BookmarkStore) { self.store = store }
    public func callAsFunction() -> [NewsArticle] { store.load() }
}

/// Toggles an article's bookmark by `id`: removes it if already saved, otherwise inserts it
/// at the front. Persists and returns the new list.
public struct ToggleBookmarkUseCase: Sendable {
    private let store: BookmarkStore
    public init(store: BookmarkStore) { self.store = store }
    public func callAsFunction(_ article: NewsArticle) -> [NewsArticle] {
        var items = store.load()
        if let index = items.firstIndex(where: { $0.id == article.id }) {
            items.remove(at: index)
        } else {
            items.insert(article, at: 0)
        }
        store.save(items)
        return items
    }
}

File: shared/src/commonMain/kotlin/com/aptrade/shared/application/NewsUseCases.kt
package com.aptrade.shared.application

import com.aptrade.shared.domain.NewsArticle
import com.aptrade.shared.domain.NewsCategory
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlin.coroutines.cancellation.CancellationException

/** Fetches market news for a category. Failures are swallowed to an empty list (macOS
 *  `try?` parity) since news is supplementary; CancellationException always rethrows. */
class FetchMarketNews(private val repository: NewsRepository) {
    @Throws(CancellationException::class)
    suspend fun execute(category: NewsCategory): List<NewsArticle> =
        try {
            repository.marketNews(category)
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            emptyList()
        }
}

/** Fetches company-specific news for a symbol. Same swallow-to-empty-list contract as
 *  [FetchMarketNews]. */
class FetchCompanyNews(private val repository: NewsRepository) {
    @Throws(CancellationException::class)
    suspend fun execute(symbol: String): List<NewsArticle> =
        try {
            repository.companyNews(symbol)
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            emptyList()
        }
}

/** Loads the user's bookmarked articles. */
class LoadBookmarks(private val store: BookmarkStore) {
    suspend fun execute(): List<NewsArticle> = store.load()
}

/** Toggles an article's bookmark state, persists the result, and returns it: present
 *  (by id) -> removed; absent -> inserted at index 0.
 *
 *  The read-modify-write runs under a [Mutex]: the current list is re-loaded from the store
 *  *inside* the lock (never taken from a possibly-stale caller snapshot), so a single shared
 *  instance serializes concurrent toggles — the second caller sees the first's write. This is
 *  why [AppGraph] hands the same instance to every view model. */
class ToggleBookmark(private val store: BookmarkStore) {
    private val mutex = Mutex()

    suspend fun execute(article: NewsArticle): List<NewsArticle> = mutex.withLock {
        val current = store.load()
        val result = if (current.any { it.id == article.id }) {
            current.filter { it.id != article.id }
        } else {
            listOf(article) + current
        }
        store.save(result)
        result
    }
}
